src/suction_grasp_scorer.cpp

```diff
@@ -0,0 +1,234 @@
+                                                    moveit_visual_tools::MoveItVisualToolsPtr visual_tools)
+{
+  overlap_vector.resize(grasp_data->suction_voxel_matrix_->getNumVoxels());
+
+  // These are in the object pose frame
+  Eigen::Vector3d box_bb_max(object_size.x() / 2.0, object_size.y() / 2.0, 0);
+  Eigen::Vector3d box_bb_min(-object_size.x() / 2.0, -object_size.y() / 2.0, 0);
+
+  if (visual_tools)
+  {
+    visual_tools->deleteAllMarkers();
+    visual_tools->trigger();
+    visual_tools->publishAxisLabeled(grasp_pose_tcp, "tcp");
+    Eigen::Isometry3d bbb1 = object_pose * Eigen::Translation3d(box_bb_max);
+    Eigen::Isometry3d bbb2 = object_pose * Eigen::Translation3d(box_bb_min);
+    visual_tools->publishAxisLabeled(bbb1, "bbb1", rviz_visual_tools::SMALL);
+    visual_tools->publishAxisLabeled(bbb2, "bbb2", rviz_visual_tools::SMALL);
+    visual_tools->trigger();
+  }
+
+  double visual_buffer = 0.0001;
+  double slices = 10;
+  double voxel_area = grasp_data->suction_voxel_matrix_->getVoxelArea();
+  Eigen::Isometry3d grasp_pose_tcp_in_box_frame = object_pose.inverse() * grasp_pose_tcp;
+  for (std::size_t voxel_id = 0; voxel_id < grasp_data->suction_voxel_matrix_->getNumVoxels(); ++voxel_id)
+  {
+    std::shared_ptr<SuctionVoxel> voxel;
+    if (!grasp_data->suction_voxel_matrix_->getSuctionVoxel(voxel_id, voxel))
+    {
+      ROS_ERROR_STREAM_NAMED("grasp_scorer.voxels", "Invalid voxel id: " << voxel_id);
+      overlap_vector.clear();
+      overlap_vector.resize(grasp_data->suction_voxel_matrix_->getNumVoxels());
+      return 0;
+    }
+
+    // These are also in the object pose frame
+    Eigen::Vector3d tmp_voxel_bb_1 = grasp_pose_tcp_in_box_frame * voxel->bottom_left_;
+    Eigen::Vector3d tmp_voxel_bb_2 = grasp_pose_tcp_in_box_frame * voxel->top_left_;
+    Eigen::Vector3d tmp_voxel_bb_3 = grasp_pose_tcp_in_box_frame * voxel->top_right_;
+    Eigen::Vector3d tmp_voxel_bb_4 = grasp_pose_tcp_in_box_frame * voxel->bottom_right_;
+
+    std::vector<double> m(4);
+    std::vector<double> b(4);
+    m[0] = (tmp_voxel_bb_1.y() - tmp_voxel_bb_2.y()) / (tmp_voxel_bb_1.x() - tmp_voxel_bb_2.x());
+    m[1] = (tmp_voxel_bb_2.y() - tmp_voxel_bb_3.y()) / (tmp_voxel_bb_2.x() - tmp_voxel_bb_3.x());
+    m[2] = (tmp_voxel_bb_3.y() - tmp_voxel_bb_4.y()) / (tmp_voxel_bb_3.x() - tmp_voxel_bb_4.x());
+    m[3] = (tmp_voxel_bb_4.y() - tmp_voxel_bb_1.y()) / (tmp_voxel_bb_4.x() - tmp_voxel_bb_1.x());
+    b[0] = tmp_voxel_bb_1.y() - m[0] * tmp_voxel_bb_1.x();
+    b[1] = tmp_voxel_bb_2.y() - m[1] * tmp_voxel_bb_2.x();
+    b[2] = tmp_voxel_bb_3.y() - m[2] * tmp_voxel_bb_3.x();
+    b[3] = tmp_voxel_bb_4.y() - m[3] * tmp_voxel_bb_4.x();
+
+    double max_y =
+        std::max(std::max(tmp_voxel_bb_1.y(), tmp_voxel_bb_2.y()), std::max(tmp_voxel_bb_3.y(), tmp_voxel_bb_4.y()));
+    double min_y =
+        std::min(std::min(tmp_voxel_bb_1.y(), tmp_voxel_bb_2.y()), std::min(tmp_voxel_bb_3.y(), tmp_voxel_bb_4.y()));
```

Comment:
Maybe better to just call `std::max` or `std::min` with an `initializer_list` of the 4 items rather than combining like this (assuming C++11)